musmart/melody/contour/step_contour.py

```diff
@@ -0,0 +1,217 @@
+        step_contour_length: int = _step_contour_length
+    ):
+        """Initialize StepContour with melody data.
+
+        Parameters
+        ----------
+        pitches : list[int]
+            List of pitch values
+        durations : list[float]
+            List of duration values measured in tatums
+        step_contour_length : int, optional
+            Length of the output step contour vector (default is 64)
+
+        References
+        ----------
+        [1] Müllensiefen, D. (2009). Fantastic: Feature ANalysis Technology Accessing
+        STatistics (In a Corpus): Technical Report v1.5
+
+        Examples
+        --------
+        >>> sc = StepContour([60, 62], [1.0, 1.0], step_contour_length=4)
+        >>> sc.contour
+        [60, 60, 62, 62]
+        """
+        if len(pitches) != len(durations):
+            raise ValueError(
+                f"The length of pitches (currently {len(pitches)}) must be equal to "
+                f"the length of durations (currently {len(durations)})"
+            )
+
+        self._step_contour_length = step_contour_length
+        self.contour = self._calculate_contour(pitches, durations)
+
+    def _normalize_durations(self, durations: list[float]) -> list[float]:
+        """Helper function to normalize note durations to fit within 4 bars of 4/4 time
+        (64 tatums total).
+
+        Parameters
+        ----------
+        durations : list[float]
+            List of duration values measured in tatums
+
+        Returns
+        -------
+        list[float]
+            List of normalized duration values
+
+        Examples
+        --------
+        >>> sc = StepContour([60], [1.0])
+        >>> sc._normalize_durations([2.0, 2.0])
+        [32.0, 32.0]
+        """
+        total_duration = sum(durations)
+        if total_duration == 0:
+            return durations
+
+        normalized = [
+            self._step_contour_length * (duration / total_duration)
+            for duration in durations
+        ]
+        return normalized
+
+    def _expand_to_vector(
+        self,
+        pitches: list[int],
+        normalized_durations: list[float]
+    ) -> list[int]:
+        """Helper function to create a vector of length step_contour_length by repeating
+        each pitch value proportionally to its normalized duration.
+
+        Parameters
+        ----------
+        pitches : list[int]
+            List of pitch values
+        normalized_durations : list[float]
+            List of normalized duration values (should sum to step_contour_length)
+
+        Returns
+        -------
+        list[int]
+            List of length step_contour_length containing repeated pitch values
+
+        Examples
+        --------
+        >>> sc = StepContour([60], [1.0], step_contour_length=4)
+        >>> sc._expand_to_vector([60, 62], [2.0, 2.0])
+        [60, 60, 62, 62]
+        """
+        result = []
+        for pitch, duration in zip(pitches, normalized_durations):
```

Comment:
The implementation of `_expand_to_vector` is succinct but not quite satisfying to me. The calculation of repetitions via rounding can accumulate errors, and then you need this hacky adjustment at the end that distorts the end of the melody. 

Here's an implementation that, while admittedly a bit more complex / less readable, avoids these distortions I think. It would be worth checking though.

```py
        if abs(sum(normalized_durations) - self._step_contour_length) > 1e-6:
            raise ValueError(
                f"The sum of normalized_durations ({sum(normalized_durations)}) must "
                f"be equal to the step contour length ({self._step_contour_length})"
            )
        # We interpret the output list as a vector of pitch samples taken
        # at times 0, 1, 2, ..., 63 where 63 = _step_contour_length - 1
        # and the length of the normalized melody is 64.

        output_length = self._step_contour_length
        output_pitches = [None for _ in range(output_length)]
        output_times = list(range(output_length))

        output_index = 0
        offset = 0.0

        # Iterate over the input pitches and durations
        for sounding_pitch, duration in zip(pitches, normalized_durations):
            offset += duration

            while True:
                # Step through the output list, and populate any time slots that
                # are occupied by the current note.
                if output_index >= output_length:
                    break
                output_time = output_times[output_index]
                if output_time >= offset:
                    break
                output_pitches[output_index] = sounding_pitch
                output_index += 1

        return output_pitches
```